### PR TITLE
PT-3441: Create a PrimaryEntityResolver

### PR DESCRIPTION
**Pull request by a-gvo**

### CODE REVIEW

**Comment by a-gvo** [comment]
Rebased on top of PT-3409.

File: components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+            this.repositories.values().forEach(this::addToMaps);
+        } catch (final RuntimeException ex) {
+            throw new InitializationException(ex.getMessage());
+        }
+    }
+
+    @Nullable
+    @Override
+    public PrimaryEntity resolveEntity(@Nullable final String entityId)
```
Comment (sdumitriu) [inline components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java]:
```
do arguments sanity check
extract id prefix
for each manager
  if manager's prefix is the same as the entity prefix
    get the primary entity from it
    if not null
      return
return null
```

Not sure about checking the prefix, though, we could just try to resolve the entity, but that will affect performance slightly. For example, on PhenomeCentral we have several patients starting with `F` (very old patients from before PhenomeCentral actually existed, and the `F` stands for `Forge`, the project that later became `Care4Rare`). Maybe we could do a first round with prefix checking, and if nothing is found, do another round without prefix checking.

File: components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        final String idPrefix = this.getIdPrefix(manager);
+        if (this.repositoriesByPrefix.containsKey(idPrefix)) {
+            throw new RuntimeException("PrimaryEntityManager objects must not have the same ID prefix. Duplicate "
+                + "prefix detected: " + idPrefix);
+        }
+        this.repositoriesByPrefix.put(idPrefix, manager);
+
+        final String managerType = this.getManagerType(manager);
+        if (this.repositoriesByType.containsKey(managerType)) {
+            throw new RuntimeException("PrimaryEntityManager objects must not have the same type. Duplicate type "
```
[the remark this answers is not in the document]
Comment (a-gvo) [reply components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java]: Fixed.

File: components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    /**
+     * Adds the manager to internal maps.
+     *
+     * @param manager the {@link PrimaryEntityManager}
+     */
+    private void addToMaps(@Nonnull final PrimaryEntityManager manager)
+    {
+        final String idPrefix = this.getIdPrefix(manager);
+        if (this.repositoriesByPrefix.containsKey(idPrefix)) {
+            throw new RuntimeException("PrimaryEntityManager objects must not have the same ID prefix. Duplicate "
```
[the remark this answers is not in the document]
Comment (a-gvo) [reply components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java]: Fixed.

File: components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java (hunk 1 above)
Comment (a-gvo) [reply components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java]: Fixed.

File: components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    private DocumentReferenceResolver<String> referenceParser;
+
+    /** Currently available primary entity managers, mapped by their entity ID prefix. */
+    private Map<String, PrimaryEntityManager> repositoriesByPrefix;
+
+    /** Currently available primary entity managers, mapped by their type. */
+    private Map<String, PrimaryEntityManager> repositoriesByType;
+
+    @Override
+    public void initialize() throws InitializationException
```
[the remark this answers is not in the document]
Comment (a-gvo) [reply components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java]: Fixed.

**Pull request merged**